// sigmajs-crypto-facade/build/lib/module/cryptoFacade.js
import { hmac } from "@noble/hashes/hmac";
import { pbkdf2 } from "@noble/hashes/pbkdf2";
import { sha512 } from "@noble/hashes/sha512";
import { CryptoContext } from "./cryptoContext";
import { Point, utils } from "./noble-secp256k1";
export class CryptoFacade {
    static normalizePoint(point) {
        return Point.fromHex(point.toHex());
    }
    static createCryptoContext() {
        return new CryptoContext();
    }
    static negatePoint(point) {
        return point.negate();
    }
    static isInfinityPoint(point) {
        return point.equals(Point.ZERO);
    }
    static multiplyPoint(point, scalar) {
        return point.multiply(scalar);
    }
    static addPoint(point1, point2) {
        return point1.add(point2);
    }
    static showPoint(point) {
        const x = point.x.toString(16).substring(0, 6);
        const y = point.y.toString(16).substring(0, 6);
        return `ECPoint(${x},${y},...)`;
    }
    static testBitZeroOfFieldElem(element) {
        return this._testBitOfFieldElem(element, 0);
    }
    static _testBitOfFieldElem(element, index) {
        return (element & BigInt(1 << index)) !== 0n;
    }
    static getEncodedOfFieldElem(element) {
        return utils._bigintTo32Bytes(element);
    }
    static getXCoord(point) {
        return point.x;
    }
    static getYCoord(point) {
        return point.y;
    }
    static getAffineXCoord(point) {
        return point.x;
    }
    static getAffineYCoord(point) {
        return point.y;
    }
    static hashHmacSHA512(key, data) {
        return hmac(sha512, key, data);
    }
    static generatePbkdf2Key(normalizedMnemonic, normalizedPass) {
        return pbkdf2(sha512, normalizedMnemonic, normalizedPass, { c: 2048, dkLen: 64 });
    }
    static normalizeChars(chars) {
        return chars.normalize("NFKD");
    }
    static getRandomBytes(length = 32) {
        return utils.randomBytes(length);
    }
}
