// @fleet-sdk/wallet/dist/wordlists.js
'use strict';

var english = require('@scure/bip39/wordlists/english');
var japanese = require('@scure/bip39/wordlists/japanese');
var korean = require('@scure/bip39/wordlists/korean');
var spanish = require('@scure/bip39/wordlists/spanish');
var simplifiedChinese = require('@scure/bip39/wordlists/simplified-chinese');
var traditionalChinese = require('@scure/bip39/wordlists/traditional-chinese');
var french = require('@scure/bip39/wordlists/french');
var italian = require('@scure/bip39/wordlists/italian');
var portuguese = require('@scure/bip39/wordlists/portuguese');
var czech = require('@scure/bip39/wordlists/czech');



Object.defineProperty(exports, "english", {
  enumerable: true,
  get: function () { return english.wordlist; }
});
Object.defineProperty(exports, "japanese", {
  enumerable: true,
  get: function () { return japanese.wordlist; }
});
Object.defineProperty(exports, "korean", {
  enumerable: true,
  get: function () { return korean.wordlist; }
});
Object.defineProperty(exports, "spanish", {
  enumerable: true,
  get: function () { return spanish.wordlist; }
});
Object.defineProperty(exports, "simplifiedChinese", {
  enumerable: true,
  get: function () { return simplifiedChinese.wordlist; }
});
Object.defineProperty(exports, "traditionalChinese", {
  enumerable: true,
  get: function () { return traditionalChinese.wordlist; }
});
Object.defineProperty(exports, "french", {
  enumerable: true,
  get: function () { return french.wordlist; }
});
Object.defineProperty(exports, "italian", {
  enumerable: true,
  get: function () { return italian.wordlist; }
});
Object.defineProperty(exports, "portuguese", {
  enumerable: true,
  get: function () { return portuguese.wordlist; }
});
Object.defineProperty(exports, "czech", {
  enumerable: true,
  get: function () { return czech.wordlist; }
});
