// sigmajs-crypto-facade/build/lib/main/cryptoFacade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CryptoFacade = void 0;
const hmac_1 = require("@noble/hashes/hmac");
const pbkdf2_1 = require("@noble/hashes/pbkdf2");
const sha512_1 = require("@noble/hashes/sha512");
const cryptoContext_1 = require("./cryptoContext");
const noble_secp256k1_1 = require("./noble-secp256k1");
class CryptoFacade {
    static normalizePoint(point) {
        return noble_secp256k1_1.Point.fromHex(point.toHex());
    }
    static createCryptoContext() {
        return new cryptoContext_1.CryptoContext();
    }
    static negatePoint(point) {
        return point.negate();
    }
    static isInfinityPoint(point) {
        return point.equals(noble_secp256k1_1.Point.ZERO);
    }
    static multiplyPoint(point, scalar) {
        return point.multiply(scalar);
    }
    static addPoint(point1, point2) {
        return point1.add(point2);
    }
    static showPoint(point) {
        const x = point.x.toString(16).substring(0, 6);
        const y = point.y.toString(16).substring(0, 6);
        return `ECPoint(${x},${y},...)`;
    }
    static testBitZeroOfFieldElem(element) {
        return this._testBitOfFieldElem(element, 0);
    }
    static _testBitOfFieldElem(element, index) {
        return (element & BigInt(1 << index)) !== 0n;
    }
    static getEncodedOfFieldElem(element) {
        return noble_secp256k1_1.utils._bigintTo32Bytes(element);
    }
    static getXCoord(point) {
        return point.x;
    }
    static getYCoord(point) {
        return point.y;
    }
    static getAffineXCoord(point) {
        return point.x;
    }
    static getAffineYCoord(point) {
        return point.y;
    }
    static hashHmacSHA512(key, data) {
        return (0, hmac_1.hmac)(sha512_1.sha512, key, data);
    }
    static generatePbkdf2Key(normalizedMnemonic, normalizedPass) {
        return (0, pbkdf2_1.pbkdf2)(sha512_1.sha512, normalizedMnemonic, normalizedPass, { c: 2048, dkLen: 64 });
    }
    static normalizeChars(chars) {
        return chars.normalize("NFKD");
    }
    static getRandomBytes(length = 32) {
        return noble_secp256k1_1.utils.randomBytes(length);
    }
}
exports.CryptoFacade = CryptoFacade;
